File: problem_solving_patterns/binary_search/binary_search.go
package main

import "fmt"

func main() {
	items := []int{1, 2, 9, 20, 31, 45, 63, 70, 100}
	fmt.Println(binarySearch(items, 63))
}

func binarySearch(s []int, val int) int {
	left := s[0]
	right := len(s) - 1
	middle := (left + right) / 2
	for s[middle] != val && left <= right {
		if s[middle] > val {
			right = middle - 1
		} else {
			left = middle + 1
		}
		middle = (left + right) / 2
	}
	if s[middle] == val {
		return middle
	}

	return -1
}

File: problem_solving_patterns/binary_search/binary_search.rb
def binary_search(arr, val)
  left = arr[0]
  right = arr.size - 1
  middle = ((left + right) / 2).floor
  while arr[middle] != val && left <= right
    if arr[middle] > val
      right = middle - 1
    else
      left = middle + 1
    end
    middle = ((left + right) / 2).floor
  end
  return middle if arr[middle] == val

  -1
end

puts binary_search([1, 2, 5, 7, 19, 24, 32], 7)
